url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/example/KalmanFilterOps.java

```diff
@@ -0,0 +1,121 @@
+
+package org.ejml.example;
+
+import org.ejml.data.DenseMatrix64F;
+
+import static org.ejml.ops.CommonOps.*;
+
+/**
+ * A Kalman filter is implemented by calling the generalized operations.  Much of the excessive
+ * memory creation/destruction has been reduced from the KalmanFilterSimple.  However, there
+ * is still room for improvement by using specialized algorithms directly.  The price paid
+ * for this better performance is the need to manually manage memory and the need to have a
+ * better understanding for how each of the operations works.
+ *
+ * @author Peter Abeles
+ */
+public class KalmanFilterOps implements KalmanFilter{
+
+    // kinematics description
+    private DenseMatrix64F F;
+    private DenseMatrix64F Q;
+    private DenseMatrix64F H;
+
+    // sytem state estimate
+    private DenseMatrix64F x;
+    private DenseMatrix64F P;
+
+    // these are predeclared for efficency reasons
+    private DenseMatrix64F a,b;
+    public DenseMatrix64F y,S,S_inv,c,d,e;
+    private DenseMatrix64F K;
+    public DenseMatrix64F Residual;
+
+    @Override
+    public void configure(DenseMatrix64F F, DenseMatrix64F Q, DenseMatrix64F H) {
+        this.F = F;
+        this.Q = Q;
+        this.H = H;
+
+        int dimenX = F.numCols;
+        int dimenZ = H.numRows;
+
+        a = new DenseMatrix64F(dimenX,1);
+        b = new DenseMatrix64F(dimenX,dimenX);
+        y = new DenseMatrix64F(dimenZ,1);
+        S = new DenseMatrix64F(dimenZ,dimenZ);
+        S_inv = new DenseMatrix64F(dimenZ,dimenZ);
+        c = new DenseMatrix64F(dimenZ,dimenX);
+        d = new DenseMatrix64F(dimenX,dimenZ);
+        K = new DenseMatrix64F(dimenX,dimenZ);
+        e=new DenseMatrix64F(1,2);
+        Residual=new DenseMatrix64F(1,1);
+        
+
+        x = new DenseMatrix64F(dimenX,1);
+        P = new DenseMatrix64F(dimenX,dimenX);
+    }
+
+    @Override
+    public void setState(DenseMatrix64F x, DenseMatrix64F P) {
+        this.x.set(x);
+        this.P.set(P);
+    }
+
+    @Override
+    public void predict() {
+
+        // x = F x
+        mult(F,x,a);
+        x.set(a);
+
+        // P = F P F' + Q
+        mult(F,P,b);
+        multTransB(b,F, P);
+        addEquals(P,Q);
+    }
+
+    @Override
+    public void update(DenseMatrix64F z, DenseMatrix64F R) {
+        // y = z - H x  y=residual
+        mult(H,x,y);
+        sub(z,y,y);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;z&#39; (index: 0 | type: DenseMatrix64F). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;R&#39;. The same parameter type is validated here for example: url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/decomposition/bidiagonal/BenchmarkBidiagonalDecomposition.java:41. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).